ddtrace/contrib/botocore/patch.py

```diff
@@ -52,23 +53,87 @@ def inject_trace_data_to_message_attributes(trace_data, entry):
         log.debug("skipping trace injection, max number (10) of MessageAttributes exceeded")
 
 
-def inject_trace_to_sqs_batch_message(args, span):
+def inject_trace_to_sqs_or_sns_batch_message(args, span):
     trace_data = {}
     HTTPPropagator.inject(span.context, trace_data)
     params = args[1]
 
-    for entry in params["Entries"]:
+    entries = params.get("Entries", params.get("PublishBatchRequestEntries", []))
+    for entry in entries:
         inject_trace_data_to_message_attributes(trace_data, entry)
 
 
-def inject_trace_to_sqs_message(args, span):
+def inject_trace_to_sqs_or_sns_message(args, span):
     trace_data = {}
     HTTPPropagator.inject(span.context, trace_data)
     params = args[1]
 
     inject_trace_data_to_message_attributes(trace_data, params)
 
 
+def inject_trace_to_event_bridge_detail(args, span):
+    params = args[1]
+    if "Entries" in params:
+        for entry in params["Entries"]:
+            if "Detail" in entry:
+                try:
+                    detail = json.loads(entry["Detail"])
+                    dd_context = detail.get("_datadog", {})
+                    HTTPPropagator.inject(span.context, dd_context)
+                    detail["_datadog"] = dd_context
+                    entry["Detail"] = json.dumps(detail)
+                except Exception:
+                    log.warning("Unable to parse Detail and inject span context")
+            else:
+                detail = {}
+                HTTPPropagator.inject(span.context, detail)
+                entry["Detail"] = json.dumps(detail)
+    else:
+        log.debug("Unable to inject context. The Event Bridge event had no Entries.")
+
+
+def get_kinesis_data_object(data, try_b64=True):
+
+    # check if data size will exceed max with headers
+    data_size = sys.getsizeof(data)
+    if data_size + 512 >= 1000000:
+        return None
+
+    try:
+        return json.loads(data)
+    except Exception:
+        if try_b64:
+            try:
+                return get_kinesis_data_object(base64.b64decode(data.encode("ascii")).decode("ascii"), False)
+            except Exception:
+                return None
+
+        return None
+
+
+def inject_trace_to_kinesis_stream_data(record, span):
+    if "Data" in record:
+        data = record["Data"]
+        data_obj = get_kinesis_data_object(data)
+        if data_obj:
+            dd_ctx = {}
+            HTTPPropagator.inject(span.context, dd_ctx)
+            data_obj["_datadog"] = dd_ctx
+            record["Data"] = json.dumps(data_obj)
+        else:
+            log.debug("Unable to parse kinesis streams data string")
+
+
+def inject_trace_to_kinesis_stream(args, span):
+    params = args[1]
+    if "Records" in params:
+        records = params.get("Records", [])
+        for record in records:
```

Comment:
Hmm, I'm not sure we want to inject context into every single message. I think the methodology is to just pass trace context into one message for the batch, recursively parsing every message is going to be a big performance hit on larger batch sizes

Reply:
It depends entirely on the customer use case. We're not trying to observe every payload in the entire pipeline here, just connect the traces.

Reply:
Updated so only the first kinesis record will get the trace injection.